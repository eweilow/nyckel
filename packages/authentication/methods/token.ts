import { GlobalAuthenticationConfig } from "../utils/globalConfig";
import fetch from "node-fetch";
import { verifyAndDecodeJWT, DecodedAccessToken } from "../jwt/verify";

type RequestTokenBody = ValidResponseTokenBody | ErrorResponseTokenBody;

type ValidResponseTokenBody = {
  access_token: string;
  refresh_token: string;
  id_token: string;
  token_type: "Bearer";
  expires_in: number;
};

type ErrorResponseTokenBody = {
  error: string;
  error_description: string;
};

function verifyTokenResponse(
  response: RequestTokenBody
): ValidResponseTokenBody {
  if ("error" in response) {
    throw new Error(response.error + ": " + response.error_description);
  }
  if (response.access_token == null) {
    throw new Error("Expected access_token to exist");
  }
  if (response.id_token == null) {
    throw new Error("Expected id_token to exist");
  }
  if (response.expires_in == null) {
    throw new Error("Expected expires_in to exist");
  }
  if (response.refresh_token == null) {
    throw new Error("Expected refresh_token to exist");
  }

  return response;
}

export async function requestToken(
  code: string,
  redirectUrl: string,
  config: GlobalAuthenticationConfig
) {
  const url = new URL(config.urls.token);
  const sendPostTo = url.href;

  const response = await fetch(sendPostTo, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      grant_type: "authorization_code",
      client_id: config.clientId,
      client_secret: config.clientSecret,
      code,
      redirect_uri: redirectUrl
    })
  });

  const body: RequestTokenBody = await response.json();

  const verifiedBody = verifyTokenResponse(body);

  let decodedIdToken!: DecodedAccessToken;
  let decodedAccessToken: DecodedAccessToken | null = null;
  let expiresUtcSeconds = Number.MAX_VALUE;
  try {
    decodedIdToken = await verifyAndDecodeJWT(verifiedBody.id_token, config);
    expiresUtcSeconds = decodedIdToken.exp;
  } catch (idTokenError) {
    idTokenError.message += " (id token)";
    throw idTokenError;
  }
  if (config.audience !== config.urls.userinfo) {
    // if configurated audience matches the Auth0 userinfo scope, the access token will not be a JWT
    try {
      decodedAccessToken = await verifyAndDecodeJWT(
        verifiedBody.access_token,
        config
      );
      if (decodedAccessToken.exp < expiresUtcSeconds) {
        expiresUtcSeconds = decodedAccessToken.exp;
      }
    } catch (accessTokenError) {
      accessTokenError.message += " (access token)";
      throw accessTokenError;
    }
  }

  const expiresUtcMilliseconds = expiresUtcSeconds * 1000;
  return {
    refreshToken: verifiedBody.refresh_token,
    accessToken: verifiedBody.access_token,
    idToken: verifiedBody.id_token,
    expires: expiresUtcMilliseconds
  };
}
