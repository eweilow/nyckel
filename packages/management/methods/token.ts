import {
  GlobalAuthenticationConfig,
  concatUrl,
  createRateLimiter,
  verifyAndDecodeJWT
} from "@nyckel/authentication";

import fetch from "node-fetch";

const managementTokenRateLimiter = createRateLimiter();

type RequestTokenBody = ValidResponseTokenBody | ErrorResponseTokenBody;

type ValidResponseTokenBody = {
  access_token: string;
  token_type: "Bearer";
};

type ErrorResponseTokenBody = {
  error: string;
  error_description: string;
};

function verifyTokenResponse(
  response: RequestTokenBody
): ValidResponseTokenBody {
  if ("error" in response) {
    throw new Error(response.error + ": " + response.error_description);
  }
  if (response.access_token == null) {
    throw new Error("Expected access_token to exist");
  }
  return response;
}

type FetchTokenResult = {
  accessToken: string;
  scope: Set<string>;
  expires: number;
};

const tokenId = "token";
export async function fetchManagementToken(
  config: GlobalAuthenticationConfig
): Promise<FetchTokenResult> {
  await managementTokenRateLimiter.wait(tokenId);
  const response = await fetch(config.urls.token, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({
      grant_type: "client_credentials",
      client_id: config.clientId,
      client_secret: config.clientSecret,
      audience: concatUrl(config.authorizationDomain, "/api/v2/")
    })
  });

  managementTokenRateLimiter.updateFromResponse(tokenId, response);

  if (response.status === 429) {
    return fetchManagementToken(config);
  }

  const body: RequestTokenBody = await response.json();
  const verifiedBody = verifyTokenResponse(body);

  const data = await verifyAndDecodeJWT(verifiedBody.access_token, config);

  return {
    accessToken: verifiedBody.access_token,
    expires: data.exp * 1000,
    scope: new Set(data.scope!.split(" "))
  };
}
