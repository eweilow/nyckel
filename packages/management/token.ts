import { createPromiseDebounce } from "./utils/promiseDebounce";
import { fetchManagementToken } from "./methods/token";

const tokenDebouncer = createPromiseDebounce(
  fetchManagementToken,
  data => data.accessToken,
  data => data.expires < Date.now() + 60000
);

export const getManagementToken = tokenDebouncer.get;
