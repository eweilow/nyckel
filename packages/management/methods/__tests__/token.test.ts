import { fetchManagementToken } from "../token";
import {
  GlobalAuthenticationConfig,
  createRateLimiter
} from "@nyckel/authentication";
import fetch from "node-fetch";

jest.mock("node-fetch");

jest.mock("@nyckel/authentication", () => {
  const instance = {
    ...jest.requireActual("@nyckel/authentication"),
    verifyAndDecodeJWT: jest.fn(() => {
      return {
        accessToken: "accessToken",
        exp: 1000,
        scope: "a scope"
      };
    }),
    rateLimiter: {
      wait: jest.fn(),
      updateFromResponse: jest.fn()
    },
    createRateLimiter: jest.fn(() => {
      return instance.rateLimiter;
    })
  };

  return instance;
});

const config: GlobalAuthenticationConfig = {
  audience: "param:audience",
  authorizationDomain: "param:authorizationDomain",
  clientId: "param:clientId",
  clientSecret: "param:clientSecret",
  urls: {
    authorization: "url:authorization",
    issuer: "url:issuer",
    jwks: "url:jwks",
    logout: "url:logout",
    token: "url:token",
    userinfo: "url:userinfo"
  }
};

describe("fetchManagementToken", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should wait for rate limiting correctly", async () => {
    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return {
          access_token: "access_token",
          token_type: "Bearer"
        };
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    await fetchManagementToken(config);

    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    const rateLimiter = require("@nyckel/authentication")
      .rateLimiter as ReturnType<typeof createRateLimiter>;
    expect(rateLimiter.wait).toBeCalledWith("token");
    expect(rateLimiter.wait).toBeCalledTimes(1);
  });

  it("should update rate limiting correctly", async () => {
    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return {
          access_token: "access_token",
          token_type: "Bearer"
        };
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    await fetchManagementToken(config);

    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    const rateLimiter = require("@nyckel/authentication")
      .rateLimiter as ReturnType<typeof createRateLimiter>;

    expect(rateLimiter.updateFromResponse).toBeCalledWith(
      "token",
      fetchResponse
    );
    expect(rateLimiter.updateFromResponse).toBeCalledTimes(1);
  });

  it("should re-attempt if rate limited", async () => {
    let responseInstance1 = {
      status: 429,
      json: jest.fn(() => {
        throw new Error("Rate limited!");
      })
    };
    let responseInstance2 = {
      status: 200,
      json: jest.fn(() => {
        return {
          access_token: "access_token",
          token_type: "Bearer"
        };
      })
    };
    ((fetch as any) as jest.Mock)
      .mockReturnValueOnce(Promise.resolve(responseInstance1))
      .mockReturnValueOnce(Promise.resolve(responseInstance2));

    await fetchManagementToken(config);

    const rateLimiter = require("@nyckel/authentication")
      .rateLimiter as ReturnType<typeof createRateLimiter>;

    expect(rateLimiter.updateFromResponse).toHaveBeenNthCalledWith(
      1,
      "token",
      responseInstance1
    );
    expect(rateLimiter.updateFromResponse).toHaveBeenNthCalledWith(
      2,
      "token",
      responseInstance2
    );
    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    expect(fetch).toBeCalledTimes(2);
    expect(rateLimiter.wait).toBeCalledTimes(2);
    expect(rateLimiter.updateFromResponse).toBeCalledTimes(2);
  });

  it("should fail if error", async () => {
    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return {
          error: "oh no",
          error_description: "something bad happened"
        };
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    const promise = fetchManagementToken(config);
    await expect(promise).rejects.toThrow();
    const err = await promise.catch(err => err);

    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    expect(fetch).toBeCalledTimes(1);
    expect(err.message).toMatchSnapshot();
  });

  it("should fail if access token was not returned", async () => {
    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return {
          data: "bogus",
          token_type: "Bearer"
        };
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    const promise = fetchManagementToken(config);
    await expect(promise).rejects.toThrow();
    const err = await promise.catch(err => err);

    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    expect(fetch).toBeCalledTimes(1);
    expect(err.message).toMatchSnapshot();
  });

  it("should fail if bearer type was not Bearer", async () => {
    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return {
          access_token: "token",
          token_type: "not bearer"
        };
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    const promise = fetchManagementToken(config);
    await expect(promise).rejects.toThrow();
    const err = await promise.catch(err => err);

    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    expect(fetch).toBeCalledTimes(1);
    expect(err.message).toMatchSnapshot();
  });

  it("should return correctly", async () => {
    const fetchResponse = {
      status: 200,
      json: jest.fn(() => {
        return {
          access_token: "token",
          token_type: "Bearer"
        };
      })
    };
    ((fetch as any) as jest.Mock).mockReturnValueOnce(
      Promise.resolve(fetchResponse)
    );

    const data = await fetchManagementToken(config);
    expect(((fetch as any) as jest.Mock).mock.calls).toMatchSnapshot();
    expect(data).toMatchSnapshot();
  });
});
