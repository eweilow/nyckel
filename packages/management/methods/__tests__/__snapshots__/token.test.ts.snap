// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchManagementToken should fail if access token was not returned 1`] = `
Array [
  Array [
    "url:token",
    Object {
      "body": "{\\"grant_type\\":\\"client_credentials\\",\\"client_id\\":\\"param:clientId\\",\\"client_secret\\":\\"param:clientSecret\\",\\"audience\\":\\"param:authorizationDomain\\"}",
      "headers": Object {
        "content-type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetchManagementToken should fail if access token was not returned 2`] = `"Expected access_token to exist"`;

exports[`fetchManagementToken should fail if error 1`] = `
Array [
  Array [
    "url:token",
    Object {
      "body": "{\\"grant_type\\":\\"client_credentials\\",\\"client_id\\":\\"param:clientId\\",\\"client_secret\\":\\"param:clientSecret\\",\\"audience\\":\\"param:authorizationDomain\\"}",
      "headers": Object {
        "content-type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetchManagementToken should fail if error 2`] = `"oh no: something bad happened"`;

exports[`fetchManagementToken should re-attempt if rate limited 1`] = `
Array [
  Array [
    "url:token",
    Object {
      "body": "{\\"grant_type\\":\\"client_credentials\\",\\"client_id\\":\\"param:clientId\\",\\"client_secret\\":\\"param:clientSecret\\",\\"audience\\":\\"param:authorizationDomain\\"}",
      "headers": Object {
        "content-type": "application/json",
      },
      "method": "POST",
    },
  ],
  Array [
    "url:token",
    Object {
      "body": "{\\"grant_type\\":\\"client_credentials\\",\\"client_id\\":\\"param:clientId\\",\\"client_secret\\":\\"param:clientSecret\\",\\"audience\\":\\"param:authorizationDomain\\"}",
      "headers": Object {
        "content-type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetchManagementToken should return correctly 1`] = `
Array [
  Array [
    "url:token",
    Object {
      "body": "{\\"grant_type\\":\\"client_credentials\\",\\"client_id\\":\\"param:clientId\\",\\"client_secret\\":\\"param:clientSecret\\",\\"audience\\":\\"param:authorizationDomain\\"}",
      "headers": Object {
        "content-type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetchManagementToken should return correctly 2`] = `
Object {
  "accessToken": "token",
  "expires": 1000000,
  "scope": Set {
    "a",
    "scope",
  },
}
`;

exports[`fetchManagementToken should update rate limiting correctly 1`] = `
Array [
  Array [
    "url:token",
    Object {
      "body": "{\\"grant_type\\":\\"client_credentials\\",\\"client_id\\":\\"param:clientId\\",\\"client_secret\\":\\"param:clientSecret\\",\\"audience\\":\\"param:authorizationDomain\\"}",
      "headers": Object {
        "content-type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`fetchManagementToken should wait for rate limiting correctly 1`] = `
Array [
  Array [
    "url:token",
    Object {
      "body": "{\\"grant_type\\":\\"client_credentials\\",\\"client_id\\":\\"param:clientId\\",\\"client_secret\\":\\"param:clientSecret\\",\\"audience\\":\\"param:authorizationDomain\\"}",
      "headers": Object {
        "content-type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;
